src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DirectoryHelper.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal static class DirectoryHelper
+    {
+        /// <summary>
+        /// Finds all the files in  a directory which meet the given criteria.
+        /// </summary>
+        /// <param name="workspaceDirectory">The directory to be searched.</param>
+        /// <param name="searchPattern">The pattern to apply when searching.</param>
+        /// <param name="ignoredDirectories">List of directories to skip when recursing.</param>
+        /// <param name="logger">An optional logger to report on exceptional situations such as <see cref="PathTooLongException"/>.</param>
+        /// <returns>A list of files within the given directory that meet the search criteria.</returns>
+        /// <remarks>This method is needed to avoid problematic folders such as "node_modules" which are known not to yield the desired results or may cause performance issues.</remarks>
+        internal static IEnumerable<string> GetFilteredFiles(string workspaceDirectory, string searchPattern, IReadOnlyCollection<string> ignoredDirectories, ILogger? logger = null)
+        {
+            if (workspaceDirectory is null)
+            {
+                throw new ArgumentNullException(nameof(workspaceDirectory));
+            }
+
+            if (searchPattern is null)
+            {
+                throw new ArgumentNullException(nameof(searchPattern));
+            }
+
+            if (ignoredDirectories is null || ignoredDirectories.Count == 0)
+            {
+                throw new ArgumentNullException(nameof(ignoredDirectories));
+            }
+
+            IEnumerable<string> files;
+            try
+            {
+                files = Directory.EnumerateFiles(workspaceDirectory, searchPattern, SearchOption.TopDirectoryOnly);
```

Comment:
So I'm pretty sure this won't work, and perhaps the best thing to do, if it meets the criteria for this issue, is just to go back to `GetFiles`. EnumerateFiles is lazily executed when the foreach loop calls `MoveNext` on the enumerator, so that is where an exception would be thrown. In order to ignore exceptions without missing any files, you'd have to not use a foreach loop, but rather manually iterate through by calling `MoveNext` and that call is the one that would need to be inside a `try..catch`. Does that make sense?

If you're not trying to create a general purpose "find a file and don't error" then it might not be worth the complexity.

Reply:
ooo good call!

Reply:
Yeah I follow, good catch.